pyiron_atomistics/vasp/base.py

```diff
@@ -1814,6 +1814,19 @@ def __del__(self):
         pass
 
 
+class VaspSpecificOptions(DataContainer):
```

Comment:
I think this should rather use `HasStorage` otherwise how you have weirdnesses like
```python
opt = VaspSpecificOptions()
opt.allow_structure_reordering = False
opt.create_group('bar')
opt.bar.allow_structure_reordering # will give True
```
This way you also don't need to worry about passing all the stuff from the `DataContainer` init.

Reply:
Okay, I will leave that bit to someone who actually understands the datacontainer objects - I will just be weeding out all the sorting dependencies from the rest of the code. It will probably take me a bit, but I'll aim to have a couple drafts up by the end of the week.

Reply:
> I think this should rather use `HasStorage` otherwise how you have weirdnesses like
> 
> ```python
> opt = VaspSpecificOptions()
> opt.allow_structure_reordering = False
> opt.create_group('bar')
> opt.bar.allow_structure_reordering # will give True
> ```
> 
> This way you also don't need to worry about passing all the stuff from the `DataContainer` init.

@pmrv Hmm, this is actually a bit of an issue with `DataContainer` IMO. `DataContainer.create_group.__doc__` asserts that it will "Add a new empty subcontainer under the given key." However, since it calls `self.__class__()` instead of `DataContainer()`, we *don't* necessarily get an empty subcontainer, but a default-initialized-sub-container. This is exactly the behaviour you demonstrate above, but I find the behaviour out of sync with both the method name and docstring. Easiest fix is the docstring, but are we sure this is the behaviour we really want for `create_group`?

I am open to using `HasStorage`, but won't this mean we need a setter and getter for each attribute to make sure it's linked to the storage, e.g.

```python
class VaspSpecificOptions(HasStorage):
        def __init__(self, group_name=None):
            super().__init__(group_name=group_name)
            self.allow_structure_reordering = True

        @property
        def allow_structure_reordering(self) -> bool:
            return self.storage.allow_structure_reordering

        @allow_structure_reordering.setter
        def allow_structure_reordering(self, allow: bool) -> None:
            self.storage.allow_structure_reordering = allow
```

I'm not totally vetoing that or anything, but it's annoying enough that I want to quickly double-check with you for alternatives. Optimal for me would be making it easy to subclass `DataContainer`. But maybe we should instead consider modifying `__setattr__` and `__getattr__` on `HasStorage` to coupe with the underlying `_storage` in this way by default? Sounds rad but risky.

Tangentially, as a mix-in, doesn't `HasStorage.__init__` need a call to `super().__init__(*args, **kwargs)` for cooperative inheritance?

Reply:
> > I think this should rather use `HasStorage` otherwise how you have weirdnesses like
> > ```python
> > opt = VaspSpecificOptions()
> > opt.allow_structure_reordering = False
> > opt.create_group('bar')
> > opt.bar.allow_structure_reordering # will give True
> > ```
> > 
> > This way you also don't need to worry about passing all the stuff from the `DataContainer` init.
> 
> @pmrv Hmm, this is actually a bit of an issue with `DataContainer` IMO. `DataContainer.create_group.__doc__` asserts that it will "Add a new empty subcontainer under the given key." However, since it calls `self.__class__()` instead of `DataContainer()`, we _don't_ necessarily get an empty subcontainer, but a default-initialized-sub-container. This is exactly the behaviour you demonstrate above, but I find the behaviour out of sync with both the method name and docstring. Easiest fix is the docstring, but are we sure this is the behaviour we really want for `create_group`?

I don't want to change it to something else then `self.__class__()`, because this would break other subclasses, like the Sphinx `Group`.  We can change the docstring of `create_group` to say that it creates a nested, fresh instance of the parent container.
 
> I am open to using `HasStorage`, but won't this mean we need a setter and getter for each attribute to make sure it's linked to the storage, e.g.
> [snip]
> I'm not totally vetoing that or anything, but it's annoying enough that I want to quickly double-check with you for alternatives. Optimal for me would be making it easy to subclass `DataContainer`. But maybe we should instead consider modifying `__setattr__` and `__getattr__` on `HasStorage` to coupe with the underlying `_storage` in this way by default? Sounds rad but risky.

Yes, that's how it would work, see [ScipyMinimizerInput](https://github.com/pyiron/pyiron_atomistics/blob/280817938ea0862cd8a400d328e4b3472d465fb3/pyiron_atomistics/interactive/scipy_minimizer.py#L295) where I recently did it this way.
I agree that this is a bit cumbersome.  @samwaseda had started a while ago on something to make it easier [here](https://github.com/pyiron/pyiron_base/pull/753).  That fizzled a bit out, but we could think about trying it again, I'm actually playing with this just now.

> Tangentially, as a mix-in, doesn't `HasStorage.__init__` need a call to `super().__init__(*args, **kwargs)` for cooperative inheritance?

True, good catch!

Reply:
I put something together [here](https://github.com/pyiron/pyiron_base/pull/854).

Reply:
Super! Yes, just adjusting the docstring for `create_group` and leaving `self.__class__()` is sufficient for me.

Comments on the rest in their PRs.